src/runtime/server/middleware/session/ipPinning.ts

```diff
@@ -0,0 +1,63 @@
+import { argon2id, hash, verify } from 'argon2'
+import { H3Event } from 'h3'
+import { useRuntimeConfig } from '#app'
+
+const argon2Options = {
+  // cryptographically-secure salt is generated automatically
+  type: argon2id, // resistant against GPU & tradeoff attacks
+  hashLength: 60
+}
+
+/**
+ * Get a hashed representation of the given IP address (for GDPR compliance)
+ * @param ip string|undefined The IP address to hash
+ */
+export const hashIpAddress = (ip: string | undefined): Promise<string | undefined> =>
+  !ip
+    ? Promise.resolve(undefined)
+    : hash(ip, argon2Options)
+
+/**
+ * Check that the given (raw) IP address and the hashed IP address match
+ * @param ip string|undefined The IP address to verify
+ * @param ipHash string|undefined The (hashed) IP address to test against
+ */
+export const ipAddressesMatch = (ip: string | undefined, ipHash: string | undefined): Promise<boolean> => !(ip && ipHash) ? Promise.resolve(false) : verify(ipHash, ip, argon2Options)
```

Comment:
Nit: do you know that if you mark functions as async it will always return promise for you? Also: multi-line functions are easier to read, boolean operations are easier to read when they are "expanded"
```suggestion
export const ipAddressesMatch = async (ip: string | undefined, ipHash: string | undefined): Promise<boolean> => {
    return (!ip || !ipHash) ? false : verify(ipHash, ip, argon2Options)
}
```

Reply:
ESLint config doesn't allow that